[MPS] Add `torch.fft.` support

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119670

Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions (see https://github.com/pytorch/pytorch/issues/120237 )

When compiling on MacOS13, implement `+[FakeMPSGraphFFTDescriptor descriptor]` as a redispatch to a real thing.

Fixes https://github.com/pytorch/pytorch/issues/78044

malfet committed:
Add FastFourierTransform

[ghstack-poisoned]

malfet committed:
Update on "Add FastFourierTransform"




[ghstack-poisoned]

malfet committed:
Update on "Add FastFourierTransform"




[ghstack-poisoned]

malfet committed:
Update on "Add FastFourierTransform"




[ghstack-poisoned]

malfet committed:
Update on "Add FastFourierTransform"




[ghstack-poisoned]

malfet committed:
Update on "Add FastFourierTransform"




[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/SpectralOpsUtils.h:
```diff
@@ -3,8 +3,14 @@
 #include <ATen/native/DispatchStub.h>
 
+namespace at {
+class Tensor;
```
Why do you need that?

malfet replied:
As linter wants to include SpectralOpsUtils.h first on the list, and want to fwd declare tensor. But can replace it with `#include` if needed

albanD replied:
If including TensorBase is enough I would.
Otherwise the forward declaration is fine yep

albanD reviewed (commented):
I'll let @kulinseth review the objc part.
The tolerance difference looks suspicious indeed. I don't know off the top of my head if FFT tolerance is supposed to scale with the input size but that sounds surprising.
If you do want to land this as-is, please make sure to open an issue about it and have a todo next to all the tolerance bumps so we can keep track of them.

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

kulinseth reviewed (approved):
Looks good.

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions

I.e. consider following script
```python
import torch
for size in [2**i for i in range(3, 12)]:
    x=torch.arange(size, dtype=torch.float, device="mps")
    print(f'| {size:4} | {(torch.fft.fft(x).cpu()-torch.fft.fft(x.cpu())).abs().max().item():.2} |')
```
will produce the following discrepancies
| size  | `.abs().max()` |
| ------------- | ------------- |
|    8 | 4.8e-07 |
|   16 | 9.5e-07 |
|   32 | 5.6e-05 |
|   64 | 0.00019 |
|  128 | 0.00052 |
|  256 | 0.002 |
|  512 | 0.0076 |
| 1024 | 0.031 |
| 2048 | 0.19 |


[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add `torch.fft.` support"


Increase tolerance for `ftt` ops, that warrants a further investigation as it grows larger with larger matrix dimensions (see https://github.com/pytorch/pytorch/issues/120237 )

When compiling on MacOS13, implement `+[FakeMPSGraphFFTDescriptor descriptor]` as a redispatch to a real thing.

Fixes https://github.com/pytorch/pytorch/issues/78044

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Lint and MPS tests are green"

albanD reviewed (approved):
SGTM

malfet commented:
@pytorchbot merge -f "Lint and MPS tests are green"
